### PR TITLE
Added warning to excessive long attack times

### LINKED ISSUE #47832: A way to prevent with long attack times?
I occasionally (w)ield containers as an unarmed character going into a place with a lot of enemies. I eventually mess up and attack while I have the bag, which is now very heavy, so the lost time results in me being basically stunned for a while.

I want some way to warn the player if their attack time will be extremely long. For example, you can get "Attacking with golf bag (39 items) will take a long time, are you sure you want to use it?".

Maybe attack time can be capped and attacks that try to go past this cap end up with penalties due to poor form. Rather than swinging a swag bag like a flail and throwing out your back, you would feebly push with it.

### PR DESCRIPTION
**Pull request by Jamuro-g**
#### Summary
Interface "Added warning to excessive long attack times"

#### Purpose of change
Accidently attacking with filled containers (backpacks, filled tanks) and other items usually not intended for melee, resulted in an excessive amount of turns passed while your character gets mauled.

Resolves #47832

#### Describe the solution

Added a yes/no/ignore warning whenever a player tries to attack with a weapon that would take more than 10 turns to execute.
the ignore feature is realised by storing the turn this option has been selected in a private var and comparing it to the current turn. (atm set to last 50 turns)

#### Testing

- Spawn in a 60L tank and fill it with water
- Spawn in a Zombie and attack
- You should get the warning dialog
- Select no to abort
- Select ignore to continue and not be warned again for 50 turns

### CODE REVIEW

**Review by topherclay — COMMENTED** [review]
Looks like maybe a rebase issue with the 1.7k+ and 1.7k- on this one commit?

File: src/character.h [hunk 1]
```diff
@@ -2957,6 +2957,7 @@ class Character : public Creature, public visitable
         tripoint cached_position;
         pimpl<inventory> cached_crafting_inventory;
 
+        int melee_warning_turn = 0;
```
Comment (anothersimulacrum) [inline src/character.h:2960]:
```suggestion
        time_point melee_warning_turn = calendar::turn_zero;
```

File: src/melee.cpp [hunk 2]
```diff
@@ -539,2090 +540,2078 @@ bool Character::melee_attack_abstract( Creature &t, bool allow_special,
...
-            }
-        } else if( player_character.sees( *this ) ) {
-            if( stumble_pen >= 60 ) {
-                add_msg( _( "%s misses and stumbles with the momentum." ), name );
-            } else if( stumble_pen >= 10 ) {
-                add_msg( _( "%s swings wildly and misses." ), name );
-            } else {
-                add_msg( _( "%s misses." ), name );
+    int now = to_turn<int>( calendar::turn );
+    if( is_avatar() && move_cost > 1000 && now > melee_warning_turn ) {
```
Comment (anothersimulacrum) [inline src/melee.cpp:546]:
```suggestion
    int now = to_turn<int>( calendar::turn );
    if( is_avatar() && move_cost > 1000 && calendar::turn > melee_warning_turn ) {
```

File: src/melee.cpp [hunk 3]
```diff
@@ -539,2090 +540,2078 @@ bool Character::melee_attack_abstract( Creature &t, bool allow_special,
...
+                             .option( "IGNORE", input_context::allow_all_keys )
+                             .query()
+                             .action;
+
+        if( action == "NO" ) {
+            return false;
+        }
+        if( action == "IGNORE" ) {
+            if( !melee_warning_turn || melee_warning_turn <= now ) {
+                melee_warning_turn = now + 50;
```
Comment (anothersimulacrum) [inline src/melee.cpp:565]:
```suggestion
            if( melee_warning_turn == calendar::turn_zero || melee_warning_turn <= calendar::turn ) {
                melee_warning_turn = calendar::turn + 50_turns;
```

**Comment by Jamuro-g** [comment]
> Looks like maybe a rebase issue with the 1.7k+ and 1.7k- on this one commit?

yep i am trying to resolve that issue atm

**Comment by actual-nh** [comment]
@anothersimulacrum: String Freeze (unfortunately)?

**Comment by anothersimulacrum** [comment]
Ah, I missed those strings.

File: src/melee.cpp [hunk 4]
```diff
@@ -539,9 +540,32 @@ bool Character::melee_attack_abstract( Creature &t, bool allow_special,
...
     int move_cost = attack_speed( *cur_weapon );
 
+    if( is_avatar() && move_cost > 1000 && calendar::turn > melee_warning_turn ) {
+        const auto &action = query_popup()
+                             .preferred_keyboard_mode( keyboard_mode::keycode )
+                             .context( "CANCEL_ACTIVITY_OR_IGNORE_QUERY" )
+                             .message( pgettext( "cancel_activity_or_ignore_query", "<color_light_red>%s %s</color>" ),
+                                       string_format(
+                                           _( "Attacking with your %1$s will take a long time." ),
+                                           cur_weapon->display_name() ), "Are you sure you want to continue?" )
```
Comment (Qrox) [inline src/melee.cpp:552]:
```suggestion
                             .message( _( "<color_light_red>Attacking with your %1$s will take a long time.  "
                                          "Are you sure you want to continue?</color>" ),
                                       cur_weapon->display_name() )
```

File: src/melee.cpp [hunk 5]
```diff
@@ -539,9 +540,32 @@ bool Character::melee_attack_abstract( Creature &t, bool allow_special,
         add_msg( m_bad, _( "This weapon is too unwieldy to attack with!" ) );
         return false;
     }
-
     int move_cost = attack_speed( *cur_weapon );
 
+    if( is_avatar() && move_cost > 1000 && calendar::turn > melee_warning_turn ) {
+        const auto &action = query_popup()
+                             .preferred_keyboard_mode( keyboard_mode::keycode )
```
Comment (Qrox) [inline src/melee.cpp:547]:
```suggestion
```
`keyboard_mode::keycode` is the default value.

File: src/melee.cpp [hunk 6]
```diff
@@ -539,9 +540,32 @@ bool Character::melee_attack_abstract( Creature &t, bool allow_special,
...
+                             .context( "CANCEL_ACTIVITY_OR_IGNORE_QUERY" )
+                             .message( pgettext( "cancel_activity_or_ignore_query", "<color_light_red>%s %s</color>" ),
+                                       string_format(
+                                           _( "Attacking with your %1$s will take a long time." ),
+                                           cur_weapon->display_name() ), "Are you sure you want to continue?" )
+                             .option( "YES", input_context::allow_all_keys )
+                             .option( "NO", input_context::allow_all_keys )
+                             .option( "IGNORE", input_context::allow_all_keys )
```
Comment (Qrox) [inline src/melee.cpp:555]:
```suggestion
                             .option( "YES" )
                             .option( "NO" )
                             .option( "IGNORE" )
```
No need for `input_context::allow_all_keys`.

**Pull request merged**